### ISSUE TITLE
What's the equivalent of `bash <(curl -fL get.awesome.io)` ?

### ISSUE THREAD

**Issue by coolaj86** [fish-shell/fish-shell#5181]
For the life of me I can't figure out what the search terms are for this.

How to use fish to "curl | bash" without pipe?

```
$ bash <(curl -fL get.example.io)
fish: Invalid redirection target: (curl -fsL get.example.io)
```

Although `curl -fL get.example.io | bash` works in every shell there's an issue with `read` (and even `exec 3<>/dev/tty; read -u 3`) in certain situations where the default `tty` cannot be accessed by a subprocess for whatever strange reason, but I need to ask the user questions as part of the install process.

Hence, my option is to show something like `curl -fL get.example.io > installer.sh; bash installer.sh` (far too difficult for today's generation of lazy and hyper experience-sensitive programmers) or `bash <(curl -fL get.example.io)`.

However, I'd also like to show fish users something that will work for them.

Of course, it would be great if fish would also support the bash way of doing it too...

**Comment by faho** [comment]
```fish
bash (curl -fL get.example.io | psub)
```

**Comment by coolaj86** [comment]
Thanks!

Would it be reasonable to consider supporting the way bash does it in the future? Or is this something that fish intentionally handles differently? Is there any rationale?

**Comment by faho** [comment]
>Would it be reasonable to consider supporting the way bash does it in the future?

No. In general, if you want 100% compatibility with bash/POSIX syntax, use a different shell.

> Is there any rationale?

From [our design docs](http://fishshell.com/docs/current/design.html):

> The law of orthogonality

>The shell language should have a small set of orthogonal features. Any situation where two features are related but not identical, one of them should be removed, and the other should be made powerful and general enough to handle all common use cases of either feature.

>Rationale:

>Related features make the language larger, which makes it harder to learn. It also increases the size of the source code, making the program harder to maintain and update.

[...]

>Subshells, command substitution and process substitution are strongly related. fish only supports command substitution, the others can be achieved either using a block or the psub shellscript function.

----


>this answer should be added to the fish docs in some place.

Huh, I could have sworn it was in the tutorial at least, but that seems to not be the case! Yes, it should be in the docs!

**Comment by vcunat** [comment]
I have quite a problem with fish's `psub`: unusability for long files (never-ending streams in my case).  This works perfectly with bash:
```bash
processing_command <(endless_stream_generator 1) <(endless_stream_generator 2)
```
but `psub` seems of no help, as the default `--file` mode inherently can't take it and surprisingly `--fifo` won't either
> You should only use this if the command produces no more than 8 KiB of output. The limit on the amount of data a FIFO can buffer varies with the OS but is typically 8 KiB, 16 KiB or 64 KiB. If you use this option and the command on the left of the psub pipeline produces more output a deadlock is likely to occur.

Am I missing some other way of doing this?

**Comment by faho** [comment]
@vcunat: #1040.

**Comment by tommyknows** [comment]
When searching for "process substitution" in the fish docs, the `psub` command comes up at the top of the list, so this is great! 

I wonder however - because I always have to google it again: would it be possible to improve the error messages on bash-style process substitutions?
For example, when running the following with a fish v3 version:
```
$ diff <(sort a.txt) <(sort b.txt)
fish: Invalid redirection target:
diff <(sort a.txt) <(sort b.txt)
     ^~~~~~~~~~~~^
```

What if fish instead detected the use of process substitution, and either hints at using `psub` directly or at least points to its docs?
```
$ diff <(sort a.txt) <(sort b.txt) 
fish: Invalid redirection target. For process substitution, use the `psub` command.
diff <(sort a.txt) <(sort b.txt)
     ^~~~~~~~~~~~^
```

(I do acknowledge that getting a terse *and* useful error message can be tricky though, and I don't believe my example provides that!)

Apart from that, this issue can probably be closed? :)
